code/game/turfs/paint_overlay.dm

```diff
@@ -0,0 +1,327 @@
+/turf/proc/remove_paint_overlay(var/erase)
+	if (paint_overlay)
+		paint_overlay.remove(erase)
+
+//refreshes the paint overlay, ensuring that it matches any changes to the turf (such as its icon changing from damage).
+/turf/proc/update_paint_overlay()
+	if (paint_overlay)
+		paint_overlay.update()
+		lighting_overlay.update_overlay()
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+/datum/paint_overlay
+	var/turf/my_turf
+	var/image/overlay
+	var/list/sub_overlays = list()
+	var/wet_color = "#FFFFFF"
+	var/wet_time = 0//world.time of the last time paint was applied that covers the whole tile AND is opaque enough (200+ alpha)
+	var/wet_duration = 10 SECONDS//relatively fast-drying
+	var/wet_amount = 3//how many steps with wet shoes
+	var/list/blood_DNA = list()
+
+	var/arbitrary_overlay_limit = 32//figured not having a limit might be silly.
+
+	var/nano_paint = FALSE//if true, main_color is added to the turf's light, enabling turfs to be lit up even without a proper light source (Might have to rework that when Europa Lights come back)
+	var/main_color = "#000000"
+
+/datum/paint_overlay/New(var/turf/_turf)
+	..()
+	my_turf = _turf
+
+//We'll need to make copies of a paint overlay when moving them to crowbar'd floor tiles.
+/datum/paint_overlay/proc/Copy()
+	var/datum/paint_overlay/copy = new()
+	copy.overlay = image('icons/turf/paint_overlays_floors.dmi',my_turf,"no_paint")
+	for (var/image/lay in sub_overlays)
+		var/image/I = image(lay)
+		copy.sub_overlays += I
+	copy.wet_color = wet_color
+	copy.wet_time = wet_time
+	copy.wet_duration = wet_duration
+	copy.wet_amount = wet_amount
+	copy.blood_DNA = blood_DNA.Copy()
+
+	copy.nano_paint = nano_paint
+	copy.main_color = main_color
+	return copy
+
+//the main proc that deals with actually adding paint on the floor
+/datum/paint_overlay/proc/apply(var/_color="#FFFFFF",var/_alpha=255,var/_mask=null,var/_mask_dir=SOUTH,var/list/_blood_DNA=list(),var/_nano_paint=FALSE)
+	my_turf.overlays -= overlay
+	if (!overlay)
+		overlay = image('icons/turf/paint_overlays_floors.dmi',my_turf,"no_paint")
+		overlay.layer = PAINT_LAYER
+	if (sub_overlays.len >= arbitrary_overlay_limit)
+		_mask = null
+		_alpha = 255
+	if (!_mask && _alpha == 255)
+		overlay.overlays.len = 0//we're applying a full opaque coat of paint so let's get rid of the other overlays
+		sub_overlays.len = 0
+		blood_DNA = list()
+	if (!_mask)
+		if (_nano_paint || (nano_paint && !_nano_paint))
+			main_color = _color
+			nano_paint = _nano_paint
+			my_turf.lighting_overlay.update_overlay()
+		var/image/new_paint_layer = image(my_turf.get_paint_icon(),my_turf,my_turf.get_paint_state(), dir = my_turf.dir)
+		new_paint_layer.color = _color
+		new_paint_layer.alpha = _alpha
+		overlay.overlays += new_paint_layer
+		sub_overlays += new_paint_layer
+		if (_alpha >= 200)
```

Comment:
any particular reason this was chosen at the alpha cutoff?

Reply:
because footprints are always displayed on top of the paint layer, so a "sufficiently opaque" paint should technically cover them, and I chose that value through testing until I found something that fit.

I'll turn it into a define.